Userland/Libraries/LibJS/Runtime/PromiseConstructor.cpp

```diff
@@ -6,15 +6,131 @@
 
 #include <LibJS/Interpreter.h>
 #include <LibJS/Runtime/AbstractOperations.h>
+#include <LibJS/Runtime/Array.h>
 #include <LibJS/Runtime/Error.h>
 #include <LibJS/Runtime/FunctionObject.h>
 #include <LibJS/Runtime/GlobalObject.h>
+#include <LibJS/Runtime/IteratorOperations.h>
 #include <LibJS/Runtime/Promise.h>
+#include <LibJS/Runtime/PromiseAllResolveElementFunction.h>
 #include <LibJS/Runtime/PromiseConstructor.h>
 #include <LibJS/Runtime/PromiseReaction.h>
+#include <LibJS/Runtime/TemporaryClearException.h>
 
 namespace JS {
 
+// 27.2.4.1.1 GetPromiseResolve ( promiseConstructor ), https://tc39.es/ecma262/#sec-getpromiseresolve
+static Value get_promise_resolve(GlobalObject& global_object, Value constructor)
+{
+    VERIFY(constructor.is_constructor());
+    auto& vm = global_object.vm();
+
+    auto promise_resolve = constructor.get(global_object, vm.names.resolve);
+    if (vm.exception())
+        return {};
+    if (!promise_resolve.is_function()) {
+        vm.throw_exception<TypeError>(global_object, ErrorType::NotAFunction, promise_resolve.to_string_without_side_effects());
+        return {};
+    }
+
+    return promise_resolve;
+}
+
+// 27.2.1.1.1 IfAbruptRejectPromise ( value, capability ), https://tc39.es/ecma262/#sec-ifabruptrejectpromise
+static Optional<Value> if_abrupt_reject_promise(GlobalObject& global_object, Value, PromiseCapability const& capability)
+{
+    auto& vm = global_object.vm();
+
+    if (auto* exception = vm.exception()) {
+        vm.clear_exception();
+        vm.stop_unwind();
+
+        (void)vm.call(*capability.reject, js_undefined(), exception->value());
+        return capability.promise;
+    }
+
+    return {};
+}
+
+static bool iterator_record_is_complete(GlobalObject& global_object, Object& iterator_record)
+{
+    auto& vm = global_object.vm();
+
+    // FIXME: Create a native iterator structure with the [[Done]] internal slot. For now, temporarily clear
+    //        the exception so we can access the "done" property on the iterator object.
+    TemporaryClearException clear_exception(vm);
+    return iterator_complete(global_object, iterator_record);
+}
+
+static void set_iterator_record_complete(GlobalObject& global_object, Object& iterator_record)
+{
+    auto& vm = global_object.vm();
+
+    // FIXME: Create a native iterator structure with the [[Done]] internal slot. For now, temporarily clear
+    //        the exception so we can access the "done" property on the iterator object.
+    TemporaryClearException clear_exception(vm);
+    iterator_record.set(vm.names.done, Value(true), Object::ShouldThrowExceptions::No);
+}
+
+// 27.2.4.1.2 PerformPromiseAll ( iteratorRecord, constructor, resultCapability, promiseResolve ), https://tc39.es/ecma262/#sec-performpromiseall
+static Value perform_promise_all(GlobalObject& global_object, Object& iterator_record, Value constructor, PromiseCapability result_capability, Value promise_resolve)
```

Comment:
Is `PromiseCapability` trivially copyable? If so, we don't need to pass it by const reference above, and if not we should probably `move()` this one below.

Reply:
`PromiseCapability` is just 3 pointers, so we can copy it